### PR TITLE
Support for Persistence

### PR DESCRIPTION
**Pull request by AndreaGron**
# Description

With this PR now it is possible to create persistent VM.
The instance controller has been updated in order to react to persistent template. A datavolume is created and it allows the creation of a pvc used to import the VM images.
It has been used a virtual machine instead of a virtual machine instance because of the need to start and stop the VMI.

### CODE REVIEW

**Comment by frisso** [comment]
Please take care of writing the proper description to your PR (starting with simple things such as "." at the end of a sentence, and capital letter at the beginning).
Persistance ==> Persistence 
(a typo in the PR title looks a very bad start....)

File: operators/pkg/instance-controller/logic.go
Comment (palexster) [inline operators/pkg/instance-controller/logic.go:27]: Some questions:
1. Commented code is dead code. If we have to delete it, let's do it (Git will store the change for us).
2. Why are we deleting such ownerReference?

File: operators/pkg/instance-controller/logic.go
Comment (palexster) [inline operators/pkg/instance-controller/logic.go:55]: This part of the code should be way more commented.

File: README.md
Comment (frisso) [inline README.md:61]:
```suggestion
The Laboratory Operator (LabOperator) implements the backend logic necessary to spawn new laboratories and persistent VMs starting from a predefined template. The instruction to install and configure the LabOperator are detailed [here](operators).
```

File: operators/pkg/instance-controller/logic.go
Comment (AndreaGron) [reply operators/pkg/instance-controller/logic.go]: We have deleted commented code.
We are deleting GlobalOwnerReference in order to pass to the function SetControllerReference ( function that belong to the controller-runtime package). Moreover we have changed the owner references of ingresses, services and secrets, that now belong to VMIs. In this way, in case of non persistent VMIs nothing changes (because VMIs are still owned by instances). However in the case of persistent VMIs, when this is deleted you should not need to have secrets, ingress and services running so also those are deleted. Whenever you will create again the VMis those resources are recreated

File: README.md
Comment (AndreaGron) [reply README.md]: Thank you for the suggestion, web have updated the read.me in the last commit

File: README.md
Comment (frisso) [inline README.md:61]: @giorio94 Hasn't be renamed that operator?
Btw, the name "lab operator" appears still here (I believe by mistake): https://github.com/netgroup-polito/CrownLabs/blob/cc8f2df7f96952970c255a30d7de0d9cee34c192/operators/pkg/instance-controller/controller_test.go#L26

File: infrastructure/containerized-data-importer/README.md
Comment (frisso) [inline infrastructure/containerized-data-importer/README.md:1]:
```suggestion
# Containerized data importer
```

File: infrastructure/containerized-data-importer/README.md
Comment (frisso) [inline infrastructure/containerized-data-importer/README.md:2]: What is this component? A service? An operator? "add-on" is not so clear to me.
build Virtual Machine Disks on PVCs for Kubevirt VMs ==> use PVCs to store Virtual Machine disks to be used by Kubevirt VMs?

File: infrastructure/containerized-data-importer/README.md
Comment (frisso) [inline infrastructure/containerized-data-importer/README.md:9]: What's the difference between 'cdi-operator' and 'cdi-cr'?

File: operators/README.md
Comment (frisso) [inline operators/README.md:61]:
```suggestion
* A persistent VirtualMachine Instance based on DataVolume and the logic to access to it
```
Is a "persistent VirtualMachine Instance" a well-known concept in CrownLabs? Or did you define it? In the latter, it should be introduced somewhere.
What is a "DataVolume"? Be more verbose, or a future student would be lost.

File: operators/README.md
Comment (frisso) [inline operators/README.md:63]:
```suggestion
All those resources are bound to the LabInstance life-cycle via the [OwnerRef property](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/)
```

File: operators/README.md
Comment (frisso) [inline operators/README.md:83]: 1) this sentence is split across two lines
2) Please rephrase the sentence, because it looks convoluted.

File: infrastructure/containerized-data-importer/README.md
Comment (AndreaGron) [reply infrastructure/containerized-data-importer/README.md:2]: Cdi is the operator that control the import the vm image in the pvc.
Actually it is not very clear we will change this description

File: operators/README.md
Comment (AndreaGron) [reply operators/README.md:83]: Actually we need to remove this sentence, because now we use vm and not vmi with a different logic, so it is deprecated

File: infrastructure/containerized-data-importer/README.md
Comment (AndreaGron) [reply infrastructure/containerized-data-importer/README.md:2]: the change " use PVCs to store Virtual Machine disks to be used by Kubevirt VMs" looks good to me

File: infrastructure/containerized-data-importer/README.md
Comment (AndreaGron) [reply infrastructure/containerized-data-importer/README.md:9]: Like in kubevirt before we deploy the operator and after the cr (instance deployment request) which triggers the actual installation

Comment (AndreaGron) [reply infrastructure/containerized-data-importer/README.md:9]: we will change the description, maybe with "#Deploy cdi CR which triggers the actual installation"

File: operators/README.md
Comment (AndreaGron) [reply operators/README.md:61]: The persistent vm is a normal crownlabs vm, but is based on a datavolume that is a custom resource of the cdi which allow the creation of the pvc, that will receive the image of the vm through an importer pod

File: README.md
Comment (giorio94) [reply README.md:61]: Yes, the correct name (at least from my understanding) should be Instance Operator. However, there are still many references both in the documentation and in the code to the old name, which should be corrected.

File: infrastructure/containerized-data-importer/README.md
Comment (frisso) [reply infrastructure/containerized-data-importer/README.md:9]: The last one looks good to me.

**Comment by frisso** [comment]
Let's wait until #401 is merged, then we review this one. I fear we can create several conflicts otherwise.

**Review by giorio94 — COMMENTED** [review]
I had a first look through your PR. Overall, the logic seems mostly ok to me. I have only some reservations about the event filtering, and especially about the usage of the CreateOrUpdate function (BTW, thanks for the conversion of the other pieces of code from the legacy CreateOrUpdate function). More details are in the inline comments below

File: operators/pkg/instance-controller/controller.go
Comment (palexster) [inline operators/pkg/instance-controller/controller.go:103]:
```suggestion
	// It continues the reconcile if the vmi doesn't exist or if datavolume is not succeeded
```

File: operators/pkg/instance-controller/controller.go
Comment (AndreaGron) [reply operators/pkg/instance-controller/controller.go:184]: We have tried without this function and it works, it was useful with previous implementation so now we will delete it.
Since we will delete this function the needToReconcile filter is useless

File: operators/pkg/instance-controller/controller.go
Comment (AndreaGron) [reply operators/pkg/instance-controller/controller.go:157]: Ok, we have removed the DatavolumeName annotation and this part of code , so now to createVMenvironment we pass the instance name.
We have also removed all the suffixes from every resources even in the test

**Review by giorio94 — COMMENTED** [review]
Some comments and proposals for modifications regarding the documentation. Ping me when also the code is stable, so that we can proceed with the review of that part.

File: operators/deploy/instance-operator/templates/clusterrole.yaml
Comment (AndreaGron) [reply operators/deploy/instance-operator/templates/clusterrole.yaml]: The first one is used by the cdi, but the other two are usless in fact we have removed them

File: operators/pkg/instance-controller/logic.go
Comment (giorio94) [inline operators/pkg/instance-controller/logic.go:101]: This could be moved before the if, as present in both branches. The `vmname` variable is probably redundant

File: operators/pkg/instance-controller/logic.go
Comment (giorio94) [inline operators/pkg/instance-controller/logic.go:167]: This should not be necessary

File: operators/pkg/instance-controller/logic.go
Comment (giorio94) [inline operators/pkg/instance-controller/logic.go:183]: This comment is no longer valid

File: operators/pkg/instance-controller/controller.go
Comment (giorio94) [inline operators/pkg/instance-controller/controller.go:149]: This change should be reverted

File: operators/pkg/instance-controller/controller.go
Comment (giorio94) [inline operators/pkg/instance-controller/controller.go:164]:
```suggestion
	type InstanceStatusPatch struct {
```

File: operators/pkg/instance-controller/controller.go
Comment (giorio94) [inline operators/pkg/instance-controller/controller.go:190]:
```suggestion
	type instanceLabelsPatch struct {
```

File: operators/pkg/instance-controller/controller.go
Comment (giorio94) [inline operators/pkg/instance-controller/controller.go:202]: I would propose the following labels, to better align with the new names:
* `crownlabs.polito.it/workspace`: template.Spec.WorkspaceRef.Name (the replace should no longer be needed, as a valid resource name is also a valid label value)
* `crownlabs.polito.it/template`: template.Name (the workspace is already implied by the workspace)
* `crownlabs.polito.it/managed-by`: `instance`

**Review by giorio94 — COMMENTED** [review]
Just some final minor comments...

Please, update also the description of this PR to clarify what it is done

File: infrastructure/virtualization/README.md
Comment (giorio94) [inline infrastructure/virtualization/README.md:27]:
```suggestion
export VERSION=$(curl -s https://github.com/kubevirt/containerized-data-importer/releases/latest | grep -o "v[0-9]\.[0-9]*\.[0-9]*")
```

File: operators/pkg/instance-creation/creation.go
Comment (AndreaGron) [reply operators/pkg/instance-creation/creation.go:193]: in fact we have removed it

**Review by giorio94 — APPROVED** [review]
/lgtm

**Review by palexster — APPROVED** [review]

**Pull request merged**
